Integrates ros_msg_utils & adds basic Time

chris-smith opened the pull request:
Each of these is roughly organized into its own commit

-migrates all source code into `src/`
-moves `ros_msg_utils` into `rosnodejs` from its separate repo
-adopts `bn.js` for 64-bit integers. I found `bn.js` to be more flexible than `bignum` and has better performance.
-Adds a basic Time interface for getting current (or sim) time (`rosnodejs.Time.now()` will return sim time if appropriate), converting between `Date` and `ros time` representations.
-Use full message definition in publisher connection headers. This fixes an issue where you could not play back `rosbags` from `rosnodejs`. As part of this, I changed some of how on the fly messages are generated to share infrastructure with message generation in order to compute the full message definition. It also changes the on the fly message signature to more closely align with the generated ones - I'm open to putting some of this back. (Resolves #38)
-Adds the ability to namespace through NodeHandles (Resolves #36)

chris-smith committed:
Move source code into src/

chris-smith committed:
Move ros_msg_utils into rosnodejs

chris-smith committed:
Move to bn.js for 64bit integers

chris-smith committed:
add basic Time interface (for sim time)

chris-smith committed:
Fix message definition for ros bags

chris-smith committed:
Small fix to pub, sub errors

chris-smith committed:
Add ability to namespace through nodehandles

chris-smith committed:
Actually send message definition in publisher header

IanTheEngineer commented on src/lib/Time.js:
```diff
@@ -0,0 +1,46 @@
+  simTime = msg.clock;
+}
+
+const Time = {
```
We should create an Issue to track the fact that Sim time will not be represented in any of the logs (they will use the wall-clock time).

IanTheEngineer commented on src/utils/messageGeneration/packages.js:
```diff
@@ -198,10 +283,9 @@ exports.findPackage = function(packageName, callback) {
-  var packagePath = process.env.ROS_PACKAGE_PATH
-  var rosPackagePaths = packagePath.split(':')
-  var directories = [rosRoot].concat(rosPackagePaths);
+  var packagePath = process.env.ROS_PACKAGE_PATH;
```
Are any checks needed to report when the `ROS_PACKAGE_PATH` isn't set?

IanTheEngineer commented on src/lib/NodeHandle.js:
```diff
@@ -19,15 +19,22 @@
-      throw new Error('Unable to create NodeHandle from type ' + typeof(node));
-    }
+  constructor(node, namespace=null) {
+    // if (!(node instanceof RosNode)) {
```
Womp womp. Remove dead code please.

IanTheEngineer reviewed (approved):
Lots of solid improvements here. Thanks for the update! Be sure to grab some engine folk to double-check before merging.

IanTheEngineer commented:
@chfritz would you give this a quick review, please? These changes touch a fair amount of your code

chfritz commented:
Will do. Should be able to find the time tonight or over the weekend.

chfritz commented:
Looks good to me. But there are minor changes I had to make to the turtlesim example to make it work again with the new code. I've created a PR for that into this branch: https://github.com/chris-smith/rosjs/pull/10. Please consider merging that before merging this one into kinetic-devel.

Great work consolidating things, @chris-smith !

chfritz committed:
Make turtle example work again (#10)

- Also added a brief description to the readme on how to run the
  turtlesim example.

chris-smith commented:
Thanks for catching that! Merged!

chris-smith committed:
Address Ian's comments
